internal/persistence/mysql/mysql.go

```diff
@@ -0,0 +1,304 @@
+	ctx, cancel := context.WithTimeout(ctx, t.queryTimeout)
+	defer cancel()
+	res, err := t.ExecContext(ctx, rawSQL, placeholders...)
+	if err != nil {
+		return err
+	}
+
+	rowCount, err := res.RowsAffected()
+	if err != nil {
+		return err
+	}
+
+	if rowCount == 0 {
+		return persistence.ErrNotFound{Key: key}
+	}
+	if rowCount != 1 {
+		return persistence.ErrInvalidRowsAffected
+	}
+
+	return nil
+}
+
+func (t *mysqlQuery) List(
+	ctx context.Context,
+	prefix string,
+	opts *persistence.ListOpts,
+) (persistence.ListResult, error) {
+	ctx, cancel := context.WithTimeout(ctx, t.queryTimeout)
+	defer cancel()
+
+	query := sq.StatementBuilder.
+		Select("`key`", "value", "COUNT(*) OVER() AS full_count").
+		From("store").
+		Where("`key` LIKE ?", prefix+"%").
+		OrderBy("`key`").
+		Limit(uint64(opts.Limit)).
+		Offset(uint64(opts.Offset))
+
+	// Parse out any provided, pre-validated CEL expression & add the proper clauses to the query.
+	var err error
+	if query, err = addFilterToQuery(query, opts.Filter); err != nil {
+		return persistence.ListResult{}, fmt.Errorf("unable to add filter CEL expression to query: %w", err)
+	}
+
+	rawSQL, placeholders, err := query.ToSql()
+	if err != nil {
+		return persistence.ListResult{}, err
+	}
+	rows, err := t.QueryContext(ctx, rawSQL, placeholders...)
+	if err != nil {
+		return persistence.ListResult{}, err
+	}
+	defer rows.Close()
+	if rows.Err() != nil {
+		return persistence.ListResult{}, rows.Err()
+	}
+
+	res := persistence.ListResult{KVList: make([]persistence.KVResult, 0, opts.Limit)}
+	for rows.Next() {
+		var kvr persistence.KVResult
+		if err := rows.Scan(&kvr.Key, &kvr.Value, &res.TotalCount); err != nil {
+			return persistence.ListResult{}, err
+		}
+		res.KVList = append(res.KVList, kvr)
+	}
+
+	return res, nil
+}
+
+func addFilterToQuery(query sq.SelectBuilder, expr *exprpb.Expr) (sq.SelectBuilder, error) {
+	// No-op when no expression is provided.
+	if expr == nil {
+		return query, nil
+	}
+
+	tags, exprFunction, err := persistence.GetTagsFromExpression(expr)
+	if err != nil {
+		return query, err
+	}
+
+	queryArgs, err := persistence.GetQueryArgsFromExprConstants(tags)
+	if err != nil {
+		return query, err
+	}
+
+	// No-op when there are no tags to filter against, to treat it as if there is no filter at all.
+	if len(queryArgs) == 0 {
+		return query, nil
+	}
+
+	tagsJSON, err := json.Marshal(queryArgs)
+	if err != nil {
+		return query, err
+	}
+
+	// As the only supported field name to filter on are tags, we can simply hard-code the predicate.
+	//
+	// NOTE: The `->` & `->>` operators shall never be used, as they are not supported in MariaDB.
+	// Instead, `JSON_EXTRACT()` should be used in its place.
+	const tagsSelector = "JSON_EXTRACT(value, '$.object.tags')"
+	if exprFunction == operators.LogicalOr {
+		query = query.Where("JSON_OVERLAPS("+tagsSelector+", ?)", tagsJSON)
+	} else {
+		query = query.Where("JSON_CONTAINS("+tagsSelector+", ?, '$')", tagsJSON)
+	}
+
+	return query, nil
```

Comment:
For the record, I did confirm this is hitting the index, using `EXPLAIN ANALYZE` in MySQL 8.0.18. Example of `JSON_OVERLAPS` hitting the index below:
```
-> Window aggregate with buffering  (actual time=0.391..0.391 rows=1 loops=1)
    -> Filter: ((store.`key` like 'c/default/o/node/%') and json_overlaps(cast(json_extract(store.`value`,_utf8mb4'$.object.tags') as char(128) array),json'["tag2"]'))  (cost=0.71 rows=1) (actual time=0.138..0.383 rows=1 loops=1)
        -> Index range scan on store using tags_idx  (cost=0.71 rows=1) (actual time=0.134..0.374 rows=31 loops=1)
```